RESOURCE_NOT_FOUND error in launching a GCP VM

I met this error when launching an 8xA100 VM on GCP. The VM was initiated despite the error.

```
I 03-19 17:18:05 optimizer.py:617] == Optimizer ==
I 03-19 17:18:05 optimizer.py:628] Target: minimizing cost
I 03-19 17:18:05 optimizer.py:640] Estimated cost: $29.4 / hour
I 03-19 17:18:05 optimizer.py:640] 
I 03-19 17:18:05 optimizer.py:712] Considered resources (1 node):
I 03-19 17:18:05 optimizer.py:760] --------------------------------------------------------------------------------------------
I 03-19 17:18:05 optimizer.py:760]  CLOUD   INSTANCE        vCPUs   Mem(GB)   ACCELERATORS   REGION/ZONE   COST ($)   CHOSEN   
I 03-19 17:18:05 optimizer.py:760] --------------------------------------------------------------------------------------------
I 03-19 17:18:05 optimizer.py:760]  GCP     a2-highgpu-8g   96      680       A100:8         us-central1   29.39         ✔     
I 03-19 17:18:05 optimizer.py:760] --------------------------------------------------------------------------------------------
I 03-19 17:18:05 optimizer.py:760] 
Launching a new cluster 'woosuk-test'. Proceed? [Y/n]: y
I 03-19 17:18:09 cloud_vm_ray_backend.py:3327] Creating a new cluster: "woosuk-test" [1x GCP(a2-highgpu-8g, {'A100': 8}, disk_size=1024)].
I 03-19 17:18:09 cloud_vm_ray_backend.py:3327] Tip: to reuse an existing cluster, specify --cluster (-c). Run `sky status` to see existing clusters.
I 03-19 17:18:10 cloud_vm_ray_backend.py:1156] To view detailed progress: tail -n100 -f /Users/woosuk/sky_logs/sky-2023-03-19-17-18-05-710862/provision.log
I 03-19 17:18:11 cloud_vm_ray_backend.py:1480] Launching on GCP us-central1 (us-central1-a)
W 03-19 17:18:25 cloud_vm_ray_backend.py:621] Got return code ZONE_RESOURCE_POOL_EXHAUSTED_WITH_DETAILS in us-central1-a (message: The zone 'projects/xxx/zones/us-central1-a' does not have enough resources available to fulfill the request.  '(resource type:compute)'.)
I 03-19 17:18:35 cloud_vm_ray_backend.py:1480] Launching on GCP us-central1 (us-central1-b)
W 03-19 17:20:16 cloud_vm_ray_backend.py:621] Got return code RESOURCE_NOT_FOUND in us-central1-b (message: The resource 'projects/xxx/zones/us-central1-b/instances/ray-woosuk-test-head-570d9247-compute' was not found)
Clusters
NAME              LAUNCHED     RESOURCES                                           STATUS   AUTOSTOP  COMMAND                       
woosuk-test       41 secs ago  1x GCP(a2-highgpu-8g, {'A100': 8}, disk_size=1024)  INIT     -         sky launch --cloud gcp ll...          

Traceback (most recent call last):
...
  File "/Users/woosuk/workspace/research/sky-proj/skypilot/sky/backends/cloud_vm_ray_backend.py", line 661, in _update_blocklist_on_gcp_error
    assert False, error
AssertionError: {'code': 'RESOURCE_NOT_FOUND', 'message': "The resource 'projects/xxx/zones/us-central1-b/instances/ray-woosuk-test-head-570d9247-compute' was not found"}
```

It seems there exists a GCP error code that is not handled by SkyPilot.

--- a/sky/backends/cloud_vm_ray_backend.py
+++ b/sky/backends/cloud_vm_ray_backend.py
@@ -659,13 +659,13 @@ class RetryingVmProvisioner(object):
                         launchable_resources.copy(zone=zone.name))
                 elif code == 'RESOURCE_NOT_FOUND':
                     # https://github.com/skypilot-org/skypilot/issues/1797
-                    # The VM may be alive on console. In the inner provision
-                    # loop we have used retries to recover but failed. The
-                    # provision loop will terminate the potentially live VMs
-                    # and move onto the next zone. Since the VM may have been
-                    # provisioned in this zone, it doesn't seem right to block
-                    # the current zone.
-                    pass
+                    # In the inner provision loop we have used retries to
+                    # recover but failed. This indicates this zone is most
+                    # likely out of capacity. The provision loop will terminate
+                    # any potentially live VMs before moving onto the next
+                    # zone.
+                    self._blocked_resources.add(
+                        launchable_resources.copy(zone=zone.name))
                 else:
                     assert False, error
         elif len(httperror_str) >= 1:

--- a/sky/backends/cloud_vm_ray_backend.py
+++ b/sky/backends/cloud_vm_ray_backend.py
@@ -657,6 +657,15 @@ class RetryingVmProvisioner(object):
                     # {'code': 8, 'message': 'There is no more capacity in the zone "europe-west4-a"; you can try in another zone where Cloud TPU Nodes are offered (see https://cloud.google.com/tpu/docs/regions) [EID: 0x1bc8f9d790be9142]'} # pylint: disable=line-too-long
                     self._blocked_resources.add(
                         launchable_resources.copy(zone=zone.name))
+                elif code == 'RESOURCE_NOT_FOUND':
+                    # https://github.com/skypilot-org/skypilot/issues/1797
+                    # The VM may be alive on console. In the inner provision
+                    # loop we have used retries to recover but failed. The
+                    # provision loop will terminate the potentially live VMs
+                    # and move onto the next zone. Since the VM may have been
+                    # provisioned in this zone, it doesn't seem right to block
+                    # the current zone.
+                    pass
                 else:
                     assert False, error
         elif len(httperror_str) >= 1:
@@ -1522,6 +1531,18 @@ class RetryingVmProvisioner(object):
                         'limit \'List requests per minute\'' in stderr):
                     logger.info(
                         'Retrying due to list request rate limit exceeded.')
+                    return True
+
+                # https://github.com/skypilot-org/skypilot/issues/1797
+                # "The resource 'projects/xxx/zones/us-central1-b/instances/ray-yyy-head-<hash>-compute' was not found" # pylint: disable=line-too-long
+                pattern = (r'\'code\': \'RESOURCE_NOT_FOUND\'.*The resource'
+                           r'.*instances\/.*-compute\' was not found')
+                result = re.search(pattern, stderr)
+                if result is not None:
+                    # Retry. Unlikely will succeed if it's due to no capacity.
+                    logger.info(
+                        'Retrying due to the possibly flaky RESOURCE_NOT_FOUND '
+                        'error.')
                     return True
 
             if ('Processing file mounts' in stdout and

